lib/unifex/code_generator/base_types/list.ex

```diff
@@ -112,44 +112,57 @@ defmodule Unifex.CodeGenerator.BaseTypes.List do
       elem_name = :"#{var_name}[i]"
       len_var_name = "#{var_name}_length"
       native_type = BaseType.generate_native_type(ctx.subtype, ctx.generator)
-      subtype = ctx.subtype
-      postproc_fun = ctx.postproc_fun
-      generator = ctx.generator
+      %{subtype: subtype, postproc_fun: postproc_fun, generator: generator} = ctx
 
       ~g"""
       ({
         int res = 1;
         int type;
         int size;
+
         ei_get_type(#{arg}->buff, #{arg}->index, &type, &size);
         #{len_var_name} = (unsigned int) size;
-        if(type == ERL_LIST_EXT) {
-          res = ei_decode_list_header(#{arg}->buff, #{arg}->index, &size);
-          #{len_var_name} = (unsigned int) size;
-          #{var_name} = malloc(sizeof(#{native_type}) * #{len_var_name});
-
-          for(unsigned int i = 0; i < #{len_var_name}; i++) {
-            #{BaseType.generate_initialization(subtype, elem_name, generator)}
-          }
 
-          for(unsigned int i = 0; i < #{len_var_name}; i++) {
-            #{BaseType.generate_arg_parse(subtype, elem_name, arg, postproc_fun, generator)}
-          }
-        } else if(type == ERL_STRING_EXT) {
+        UnifexCNodeInBuff *unifex_buff;
+        unifex_buff = malloc(sizeof(UnifexCNodeInBuff));
+        if(type == ERL_STRING_EXT) {
           char *p = malloc(sizeof(char) * #{len_var_name});
           res = ei_decode_string(#{arg}->buff, #{arg}->index, p);
-          #{var_name} = malloc(sizeof(int) * #{len_var_name});
+          unsigned char *unsigned_p = (unsigned char *)p;
+
+          ei_x_buff buff;
+          ei_x_new_with_version(&buff);
+          ei_x_encode_list_header(&buff, #{len_var_name});
           for(unsigned int i = 0; i < #{len_var_name}; i++) {
-            #ifdef __CHAR_UNSIGNED
-              #{elem_name} = (int)p[i];
-            #else
-              #{elem_name} = (int)p[i];
-              if(#{elem_name} < 0) {
-                #{elem_name} = #{elem_name} + 256;
-              }
-            #endif
+            ei_x_encode_ulong(&buff, (unsigned long)unsigned_p[i]);
           }
+          ei_x_encode_empty_list(&buff);
+
+          int version;
+          int index = 0;
+          ei_decode_version(buff.buff, &index, &version);
+          res = ei_get_type(buff.buff, &index, &type, &size);
+          unifex_buff->buff = buff.buff;
+          unifex_buff->index = &index;
+        } else {
+          unifex_buff->buff = #{arg}->buff;
+          unifex_buff->index = #{arg}->index;
+        }
```

Comment:
consider moving that into something like `unifex_cnode_string_to_list`

Reply:
I moved only part that converts string to list because it seems to me a little more readable and easier to debbug than moving this whole code?